File: python/src/day09.py
#!/usr/bin/env python

from pathlib import Path
from typing import List

INPUT_FILE = Path.cwd().parent / "inputs" / "day09.txt"


def transform_input(input: str) -> List[int]:
    return [int(x) for x in input.strip().split("\n")]


def in_prev_25_sums(number: int, numbers: List[int]) -> bool:
    complements = set()

    for x in numbers:
        if x in complements:
            return True

        complements.add(number - x)

    return False


def part1(numbers: List[int]) -> int:
    for i, number in enumerate(numbers[25:]):
        if not in_prev_25_sums(number, numbers[i : i + 25]):
            return number

    raise Exception("Didn't find any weakness.")


def part2(numbers: List[int]) -> int:
    target = part1(numbers)
    contiguous_range = find_continguous_range(numbers, target)
    return min(contiguous_range) + max(contiguous_range)


def find_continguous_range(numbers: List[int], target: int) -> List[int]:
    current_sum = numbers[0]
    bottom = 0
    top = 0

    while top < len(numbers):
        if current_sum == target:
            return numbers[bottom:top]

        if current_sum > target:
            current_sum -= numbers[bottom]
            bottom += 1
        elif current_sum < target:
            top += 1
            current_sum += numbers[top]

    raise Exception("Didn't find a contiguous sum.")


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        numbers = transform_input(f.read())

    print(f"Part 1: {part1(numbers)}")
    print(f"Part 2: {part2(numbers)}")

File: rust/src/day09.rs
use std::collections::HashSet;

static INPUT: &str = include_str!("../../inputs/day09.txt");

pub fn run() {
    let numbers = transform_input(INPUT);

    println!("Part 1: {}", part1(&numbers));
    println!("Part 2: {}", part2(&numbers));
}

fn transform_input(input: &'static str) -> Vec<i64> {
    input
        .trim_end()
        .split('\n')
        .map(|x| x.parse().unwrap())
        .collect()
}

fn part1(numbers: &[i64]) -> i64 {
    *numbers
        .iter()
        .skip(25)
        .enumerate()
        .find(|&(idx, elem)| !in_25_prev_sums(*elem, &numbers[idx..idx + 25]))
        .unwrap()
        .1
}

fn in_25_prev_sums(number: i64, numbers: &[i64]) -> bool {
    let mut complements: HashSet<i64> = HashSet::new();

    for x in numbers {
        if complements.contains(x) {
            return true;
        }

        complements.insert(number - x);
    }

    false
}

fn part2(numbers: &[i64]) -> i64 {
    let target = part1(numbers);
    let contiguous_range = find_contiguous_range(numbers, target);

    let min = contiguous_range.iter().min().unwrap();
    let max = contiguous_range.iter().max().unwrap();

    min + max
}

fn find_contiguous_range(numbers: &[i64], target: i64) -> &[i64] {
    let mut current_sum = numbers[0];
    let mut bottom = 0;
    let mut top = 0;

    while top < numbers.len() {
        if current_sum == target {
            return &numbers[bottom..top];
        }

        if current_sum > target {
            current_sum -= numbers[bottom];
            bottom += 1;
        } else {
            top += 1;
            current_sum += numbers[top];
        }
    }

    panic!("No contiguous range found.");
}
